// @colony/colony-js-client/lib/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EMPTY_ADDRESS = exports.AUTHORITY_ROLES = exports.ADMIN_ROLE = exports.OWNER_ROLE = exports.DEFAULT_DOMAIN_ID = exports.ROLES = exports.MANAGER_ROLE = exports.EVALUATOR_ROLE = exports.WORKER_ROLE = undefined;

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _ROLES, _AUTHORITY_ROLES;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var WORKER_ROLE = exports.WORKER_ROLE = 'WORKER';
var EVALUATOR_ROLE = exports.EVALUATOR_ROLE = 'EVALUATOR';
var MANAGER_ROLE = exports.MANAGER_ROLE = 'MANAGER';

var ROLES = exports.ROLES = (_ROLES = {}, (0, _defineProperty3.default)(_ROLES, MANAGER_ROLE, 0), (0, _defineProperty3.default)(_ROLES, EVALUATOR_ROLE, 1), (0, _defineProperty3.default)(_ROLES, WORKER_ROLE, 2), _ROLES);

var DEFAULT_DOMAIN_ID = exports.DEFAULT_DOMAIN_ID = 1;

var OWNER_ROLE = exports.OWNER_ROLE = 'OWNER';
var ADMIN_ROLE = exports.ADMIN_ROLE = 'ADMIN';

var AUTHORITY_ROLES = exports.AUTHORITY_ROLES = (_AUTHORITY_ROLES = {}, (0, _defineProperty3.default)(_AUTHORITY_ROLES, OWNER_ROLE, 0), (0, _defineProperty3.default)(_AUTHORITY_ROLES, ADMIN_ROLE, 1), _AUTHORITY_ROLES);

var EMPTY_ADDRESS = exports.EMPTY_ADDRESS = '0x0000000000000000000000000000000000000000';
