// @colony/colony-js-client/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WORKER_ROLE = exports.ROLES = exports.OWNER_ROLE = exports.MANAGER_ROLE = exports.EVALUATOR_ROLE = exports.EMPTY_ADDRESS = exports.AUTHORITY_ROLES = exports.ADMIN_ROLE = undefined;

var _constants = require('./constants');

Object.defineProperty(exports, 'ADMIN_ROLE', {
  enumerable: true,
  get: function get() {
    return _constants.ADMIN_ROLE;
  }
});
Object.defineProperty(exports, 'AUTHORITY_ROLES', {
  enumerable: true,
  get: function get() {
    return _constants.AUTHORITY_ROLES;
  }
});
Object.defineProperty(exports, 'EMPTY_ADDRESS', {
  enumerable: true,
  get: function get() {
    return _constants.EMPTY_ADDRESS;
  }
});
Object.defineProperty(exports, 'EVALUATOR_ROLE', {
  enumerable: true,
  get: function get() {
    return _constants.EVALUATOR_ROLE;
  }
});
Object.defineProperty(exports, 'MANAGER_ROLE', {
  enumerable: true,
  get: function get() {
    return _constants.MANAGER_ROLE;
  }
});
Object.defineProperty(exports, 'OWNER_ROLE', {
  enumerable: true,
  get: function get() {
    return _constants.OWNER_ROLE;
  }
});
Object.defineProperty(exports, 'ROLES', {
  enumerable: true,
  get: function get() {
    return _constants.ROLES;
  }
});
Object.defineProperty(exports, 'WORKER_ROLE', {
  enumerable: true,
  get: function get() {
    return _constants.WORKER_ROLE;
  }
});

require('./paramTypes');

var _index = require('./ColonyNetworkClient/index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = _index2.default;
